SimPEG/EM/NSEM/FieldsNSEM.py

```diff
@@ -0,0 +1,563 @@
+from SimPEG import Survey, Utils, Problem, np, sp, mkvc
+from scipy.constants import mu_0
+import numpy as np
+import scipy.sparse as sp
+from SimPEG.Utils import Zero, Identity
+from SimPEG.EM.Utils import omega
+
+
+##############
+### Fields ###
+##############
+class BaseNSEMFields(Problem.Fields):
+    """Field Storage for a NSEM method."""
+    knownFields = {}
+    dtype = complex
+
+###########
+# 1D Fields
+###########
+class Fields1D_ePrimSec(BaseNSEMFields):
+    """
+    Fields storage for the 1D NSEM solution.
+
+    Solving for e fields, using primary/secondary formulation
+    """
+    knownFields = {'e_1dSolution':'F'}
+    aliasFields = {
+                    'e_1d' : ['e_1dSolution','F','_e'],
+                    'e_1dPrimary' : ['e_1dSolution','F','_ePrimary'],
+                    'e_1dSecondary' : ['e_1dSolution','F','_eSecondary'],
+                    'b_1d' : ['e_1dSolution','E','_b'],
+                    'b_1dPrimary' : ['e_1dSolution','E','_bPrimary'],
+                    'b_1dSecondary' : ['e_1dSolution','E','_bSecondary']
+                  }
+
+    def __init__(self,mesh,survey,**kwargs):
+        BaseNSEMFields.__init__(self,mesh,survey,**kwargs)
+
+    def _ePrimary(self, eSolution, srcList):
+        """
+        Primary electric field from source
+
+        :param numpy.ndarray eSolution: field we solved for
+        :param list srcList: list of sources
+        :rtype: numpy.ndarray
+        :return: primary electric field as defined by the sources
+        """
+        ePrimary = np.zeros_like(eSolution)
+        for i, src in enumerate(srcList):
+            ep = src.ePrimary(self.survey.prob)
+            if ep is not None:
+                ePrimary[:,i] = ep[:,-1]
+        return ePrimary
+
+    def _eSecondary(self, eSolution, srcList):
+        """
+        Secondary electric field is the thing we solved for
+
+        :param numpy.ndarray eSolution: field we solved for
+        :param list srcList: list of sources
+        :rtype: numpy.ndarray
+        :return: secondary electric field
+        """
+        return eSolution
+
+    # Overwriting a base FDEM method, could use it.
+    def _e(self, eSolution, srcList):
+        """
+        Total electric field is sum of primary and secondary
+
+        :param numpy.ndarray solution: field we solved for
+        :param list srcList: list of sources
+        :rtype: numpy.ndarray
+        :return: total electric field
+        """
+        return self._ePrimary(eSolution,srcList) + self._eSecondary(eSolution,srcList)
+
+    def _eDeriv(self, src, v, adjoint = False):
```

Comment:
this will need to take `du_dm_v` and `v`. What you are calling `v` at the moment is actually`du_dm_v`. If you switch formulations so that you do have an eDeriv_m, this will show up as a bug.